### PR TITLE
feat: TypeScript support with JOI validation

### PR DESCRIPTION
**Pull request by lukehesluke**
Examples in a comment below

4 GH issues that need to be raised in order of importance

1. Model inheritance for TS. Event should perhaps be a union of all models which inherit Event - SessionSeries, ScheduledSession, etc.

    🚨  Until this is done, validating e.g. a ScheduledSession with `superEvent` as a SessionSeries will fail, as Joi will say "this is not an Event". See tmp/src/examples/event-inheritance-validation-examples.ts
2. Upgrade to latest JOI version if/when they support something that can replace `.lazy(..)` for allowing mutual recursion

    ⚠️  Until this is done, users will not be able to use the latest version of JOI
3. Enum inheritance. Allow `MeasurementTypeEnumeration` to take on all the values of its enumeration sub-types
4. Why is EventStatusType being generated as a model as well as an enum? Just seems like a weird little bug. Hacked around in this PR

----

The models-ts code is in `/tmp` for now. I just included it here to ease my own development and maybe ease the review (?). However, once approved, I'll:

1. move them into a models-ts PR along with all the generated files
2. In that PR, also enable pre-commit testing with husky. This isn't possible in this PR yet for reasons

### CODE REVIEW

**Comment by lukehesluke** [comment]
Examples:

An enum:

```ts
// This file was generated
import * as Joi from 'joi';
import { OaValidationError } from '../oaValidationError';

/**
 * RequiredStatusType
 */
export type Type = 
  
  | 'https://openactive.io/Required'
  | 'https://openactive.io/Optional'
  | 'https://openactive.io/Unavailable'
  ;

/**
 * RequiredStatusType - Validation schema (w/ JOI)
 */
export const JoiSchema = Joi.string().valid(
  'https://openactive.io/Required',
  'https://openactive.io/Optional',
  'https://openactive.io/Unavailable',
);

/**
 * Runtime validator for RequiredStatusType.
 *
 * If some data has one of the values of the RequiredStatusType enum, it will be returned with the correct type.
 * Otherwise, this function returns an `OAValidationError` with details about why the data does not match.
 *
 * Use this to e.g. check a JSON object received in an HTTP request. Example usage (for an express request handler):
 *
 * ```ts
 * const maybeRequiredStatusType = RequiredStatusType.validate(req.body); // `req.body` will have type `any` or `unknown`
 * if (maybeRequiredStatusType instanceof OaValidationError) {
 *   // From this point on, `maybeRequiredStatusType` will have type `OaValidationError`
 *   const error = maybeRequiredStatusType;
 *   // Do something with the error. Maybe ignore it? Or log it? Or throw? Up to you.
 * }
 * // From this point on, `maybeRequiredStatusType` will have type `RequiredStatusType.Type`
 * const requiredStatusType = maybeRequiredStatusType;
 * ```
 */
export function validate(maybeRequiredStatusType: unknown): Type | OaValidationError {
  const { value, error } = JoiSchema.validate(maybeRequiredStatusType);
  if (error) {
    return new OaValidationError('RequiredStatusType', maybeRequiredStatusType, error);
  }
  /* Joi does not implement TS Type Guards, so TS does not implicitly know that this has now been validated
  to have the right type. Therefore, we just cast it to `any`. */
  return value as any;
}
```

A model:

```ts
// This file was generated
import * as Joi from 'joi';
import { OaValidationError } from '../oaValidationError';
import { oa, schema } from '..';

/**
 * Event
 *
 * This type is derived from https://schema.org/Event, which means that any of this type's properties within schema.org may also be used.
 */
export type Type = {
  '@type': 'Event';
  /**
   * A local non-URI identifier for the resource
   *
   * ```json
   * "identifier": "SB1234"
   * ```
   */
  identifier?: number | string | oa.PropertyValue.Type | oa.PropertyValue.Type[];
  /**
   * The name of the event
   *
   * ```json
   * "name": "Speedball"
   * ```
   */
  name?: string;
  /**
   * A plain text description of the event, which must not include HTML or other markup.
   *
   * ```json
   * "description": "A fast paced game that incorporates netball, handball and football."
   * ```
   */
  description?: string;
  /**
   * A unique url based identifier for the record
   *
   * ```json
   * "@id": "https://example.com/event/1234512345"
   * ```
   */
  '@id'?: string;
  /**
   * Provide additional, specific documentation for participants about how disabilities are, or can be supported at the Event.
   *
   * ```json
   * "accessibilityInformation": "This route has been British Cycling assessed as an accessible route, meaning it is suitable for the majority of adaptive bikes. The route will have no or low levels of traffic, there will be plenty of space and will have a good surface throughout. If you have any questions about using this route on an adaptive bike on this ride, please use visit https://www.letsride.co.uk/accessibility or call 0123 456 7000 and ask for the Recreation team."
   * ```
   */
  accessibilityInformation?: string;
  /**
   * Used to specify the types of disabilities or impairments that are supported at an event.
   *
   * ```json
   * "accessibilitySupport": [
   *   {
   *     "@type": "Concept",
   *     "@id": "https://openactive.io/accessibility-support#1393f2dc-3fcc-4be9-a99f-f1e51f5ad277",
   *     "prefLabel": "Visual impairment",
   *     "inScheme": "https://openactive.io/accessibility-support"
   *   }
   * ]
   * ```
   */
  accessibilitySupport?: oa.Concept.Type[];
  /**
   * Specifies the physical activity or activities that will take place during an event.
   *
   * ```json
   * "activity": [
   *   {
   *     "@type": "Concept",
   *     "@id": "https://openactive.io/activity-list#5e78bcbe-36db-425a-9064-bf96d09cc351",
   *     "prefLabel": "Bodypump™",
   *     "inScheme": "https://openactive.io/activity-list"
   *   }
   * ]
   * ```
   */
  activity?: oa.Concept.Type[];
  /**
   * Free text restrictions that must be displayed prominently to the user before booking. This property must only contain restrictions not described by `oa:ageRestriction` or `oa:genderRestriction`.
   *
   * ```json
   * "additionalAdmissionRestriction": [
   *   "Participants younger than 12 must be accompanied by an adult",
   *   "Participants must be comfortable standing for long periods of time"
   * ]
   * ```
   */
  additionalAdmissionRestriction?: string[];
  /**
   * Indicates that an event is recommended as being suitable for or is targetted at a specific age range.
   *
   * ```json
   * "ageRange": {
   *   "@type": "QuantitativeValue",
   *   "minValue": 50,
   *   "maxValue": 60
   * }
   * ```
   */
  ageRange?: oa.QuantitativeValue.Type;
  /**
   * The enforced attendee age range requirement of the Event or Offer, that must be displayed prominently to the user before booking.
   *
   * ```json
   * "ageRestriction": {
   *   "@type": "QuantitativeValue",
   *   "minValue": 15,
   *   "maxValue": 60
   * }
   * ```
   */
  ageRestriction?: oa.QuantitativeValue.Type;
  /**
   * Provides additional notes and instructions for event attendees, for example more information on how to find the event, what to bring, etc. The value of this property must not include HTML or other markup.
   *
   * ```json
   * "attendeeInstructions": "Ensure you bring trainers and a bottle of water."
   * ```
   */
  attendeeInstructions?: string;
  /**
   * Provides a set of tags that help categorise and describe an event, e.g. its intensity, purpose, etc.
   *
   * ```json
   * "category": [
   *   "High Intensity"
   * ]
   * ```
   */
  category?: string[] | oa.Concept.Type[];
  /**
   * A Person who contributes to the facilitation of the Event.
   *
   * ```json
   * "contributor": [
   *   {
   *     "@type": "Person",
   *     "familyName": "Smith",
   *     "givenName": "Nicole",
   *     "@id": "https://example.com/locations/1234ABCD/leaders/89",
   *     "identifier": 89
   *   }
   * ]
   * ```
   */
  contributor?: oa.Person.Type[];
  /**
   * The duration of the event given in [ISO8601] format.
   *
   * ```json
   * "duration": "PT1H"
   * ```
   */
  duration?: string;
  /**
   * The eventAttendanceMode of an event indicates whether it occurs online, offline, or a mix.
   *
   * ```json
   * "eventAttendanceMode": "https://schema.org/OnlineEventAttendanceMode"
   * ```
   */
  eventAttendanceMode?: schema.EventAttendanceModeEnumeration.Type;
  /**
   * The status of an event. Can be used to indicate rescheduled or cancelled events
   *
   * ```json
   * "eventStatus": "https://schema.org/EventScheduled"
   * ```
   */
  eventStatus?: schema.EventStatusType.Type;
  /**
   * Indicates that an event is restricted to male, female or a mixed audience. This information must be displayed prominently to the user before booking. If a gender restriction isn't specified then applications should assume that an event is suitable for a mixed audience.
   *
   * ```json
   * "genderRestriction": "https://openactive.io/FemaleOnly"
   * ```
   */
  genderRestriction?: oa.GenderRestrictionType.Type;
  /**
   * An image or photo that depicts the event, e.g. a photo taken at a previous event.
   *
   * ```json
   * "image": [
   *   {
   *     "@type": "ImageObject",
   *     "url": "http://example.com/static/image/speedball_large.jpg",
   *     "thumbnail": [
   *       {
   *         "@type": "ImageObject",
   *         "url": "http://example.com/static/image/speedball_thumbnail.jpg"
   *       }
   *     ]
   *   }
   * ]
   * ```
   */
  image?: oa.ImageObject.Type[];
  /**
   * Whether the Event is accessible without charge.
   *
   * ```json
   * "isAccessibleForFree": "true"
   * ```
   */
  isAccessibleForFree?: boolean;
  /**
   * A boolean property that indicates whether an Event will be coached. This flag allows an Event to be marked as being coached without having to specify a named individual as a coach. This addresses both privacy concerns and also scenarios where the actual coach may only be decided on the day.
   *
   * ```json
   * "isCoached": "true"
   * ```
   */
  isCoached?: boolean;
  /**
   * Refers to a person (schema:Person) who will be leading an event. E.g. a coach. This is a more specific role than an organiser or a contributor. The person will need to have given their consent for their personal information to be present in the Open Data.
   *
   * ```json
   * "leader": [
   *   {
   *     "@type": "Person",
   *     "familyName": "Smith",
   *     "givenName": "Nicole",
   *     "gender": "https://schema.org/Male",
   *     "@id": "https://example.com/locations/1234ABCD/leaders/89",
   *     "identifier": 89
   *   }
   * ]
   * ```
   */
  leader?: oa.Person.Type[];
  /**
   * A general purpose property for specifying the suitability of an event for different participant “levels”. E.g. `Beginner`, `Intermediate`, `Advanced`. Or in the case of martial arts, specific belt requirements.
   *
   * ```json
   * "level": [
   *   "Beginner"
   * ]
   * ```
   */
  level?: string[] | oa.Concept.Type[];
  /**
   * The location at which the event will take place. Or, in the case of events that may span multiple locations, the initial meeting or starting point.
   * It is recommended that locations should be specified as a [Place](/models/place) complete with a fully described geographic location and/or address.
   * If only an address is available then this should be described as a [PostalAddress](/models/postaladdress).
   * Applications may use [schema:Text](https://schema.org/Text) to provide a more general description of a location ("In Victoria Park, near the lake"), but this is not recommended: consuming applications will be unable to help users discover opportunities based on their location.
   *
   * ```json
   * "location": {
   *   "@type": "Place",
   *   "@id": "https://example.com/locations/1234ABCD",
   *   "identifier": "1234ABCD",
   *   "address": {
   *     "addressLocality": "New Malden",
   *     "addressRegion": "London",
   *     "postalCode": "NW5 3DU",
   *     "streetAddress": "Raynes Park High School, 46A West Barnes Lane",
   *     "@type": "PostalAddress"
   *   },
   *   "description": "Raynes Park High School in London",
   *   "geo": {
   *     "latitude": 51.4034423828125,
   *     "longitude": -0.2369088977575302,
   *     "@type": "GeoCoordinates"
   *   },
   *   "name": "Raynes Park High School",
   *   "telephone": "01253 473934"
   * }
   * ```
   */
  location?: oa.Place.Type;
  /**
   * The maximum capacity of the Event.
   *
   * ```json
   * "maximumAttendeeCapacity": 30
   * ```
   */
  maximumAttendeeCapacity?: number;
  /**
   * Indicates the maximum number of connections to a shared virtual space.
   *
   * ```json
   * "maximumVirtualAttendeeCapacity": 20
   * ```
   */
  maximumVirtualAttendeeCapacity?: number;
  /**
   * Instructions for the attendees of an Event about where they should meet the organizer or leader at the start of the event. Some larger locations may have several possible meeting points, so this property provides additional more specific directions.
   *
   * ```json
   * "meetingPoint": "At the entrance to the park"
   * ```
   */
  meetingPoint?: string;
  /**
   * An array of schema:Offer that include the price of attending.
   *
   * ```json
   * "offers": [
   *   {
   *     "@type": "Offer",
   *     "identifier": "OX-AD",
   *     "name": "Adult",
   *     "price": 3.3,
   *     "priceCurrency": "GBP",
   *     "url": "https://profile.everyoneactive.com/booking?Site=0140&Activities=1402CBP20150217&Culture=en-GB"
   *   }
   * ]
   * ```
   */
  offers?: oa.Offer.Type[];
  /**
   * The person or organization ultimately responsible for an event. An organizer might be an  schema:Organization or a schema:Person.
   *
   * ```json
   * "organizer": {
   *   "@type": "Organization",
   *   "name": "Central Speedball Association",
   *   "url": "http://www.speedball-world.com"
   * }
   * ```
   */
  organizer?: oa.Person.Type | oa.Organization.Type;
  /**
   * Indicates that an event will be organised according to a specific Programme.
   *
   * ```json
   * "programme": {
   *   "@type": "Brand",
   *   "name": "Play Ball!",
   *   "url": "http://example.org/brand/play-ball"
   * }
   * ```
   */
  programme?: oa.Brand.Type;
  /**
   * The number of places that are still available for the Event.
   *
   * ```json
   * "remainingAttendeeCapacity": 20
   * ```
   */
  remainingAttendeeCapacity?: number;
  /**
   * Provides a note from an organizer relating to how this Event is scheduled.
   *
   * ```json
   * "schedulingNote": "This event doesn't run during school holidays"
   * ```
   */
  schedulingNote?: string;
  /**
   * The start date and time of the event. Can be specified as a schema:Date or schema:DateTime.
   *
   * ```json
   * "startDate": "2018-01-27T12:00:00Z"
   * ```
   */
  startDate?: string;
  /**
   * The end date and time of the event. Can be specified as a schema:Date or  schema:DateTime
   * It is recommended that publishers provide either an schema:endDate or a schema:duration for an event.
   *
   * ```json
   * "endDate": "2018-01-27T12:00:00Z"
   * ```
   */
  endDate?: string;
  /**
   * Relates a parent event to a child event. Properties describing the parent event can be assumed to apply to the child, unless otherwise specified. A child event might be a specific instance of an Event within a schedule
   */
  subEvent?: oa.Event.Type[];
  /**
   * Relates a child event to a parent event. Properties describing the parent event can be assumed to apply to the child, unless otherwise specified. A parent event might specify a recurring schedule, of which the child event is one specific instance
   */
  superEvent?: oa.Event.Type;
  /**
   * A URL to a web page (or section of a page) that describes the event.
   *
   * ```json
   * "url": "https://example.com/event/1234"
   * ```
   */
  url?: string;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Sometimes a description is stored with formatting (e.g. href, bold, italics, embedded YouTube videos). This formatting can be useful for data consumers. This property must contain HTML.
   * 
   * If you are using this property, please join the discussion at proposal [#276](https://github.com/openactive/modelling-opportunity-data/issues/276).
   */
  formattedDescription?: string;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * The distance of a run, cycle or other activity. Must also include units.
   * 
   * If you are using this property, please join the discussion at proposal [#275](https://github.com/openactive/modelling-opportunity-data/issues/275).
   */
  distance?: oa.QuantitativeValue.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * For events that have an unlimited number of tickets, captures the number of attendees (actual attendance).
   * 
   * If you are using this property, please join the discussion at proposal [#274](https://github.com/openactive/modelling-opportunity-data/issues/274).
   */
  attendeeCount?: number;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * For events that have an unlimited number of tickets, captures the number of registrations (intention to attend).
   * 
   * If you are using this property, please join the discussion at proposal [#273](https://github.com/openactive/modelling-opportunity-data/issues/273).
   */
  registrationCount?: number;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * A property that details whether the event is suitable for wheelchair access. Placed on Event as this field could be used to detail whether the Event is suitable, as well as the Place.
   * 
   * If you are using this property, please join the discussion at proposal [#166](https://github.com/openactive/modelling-opportunity-data/issues/166).
   */
  isWheelchairAccessible?: boolean;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * A property that allows an Event duration to be represented as a range (e.g. 0-30mins, 30-60mins, 60-90mins, 90+).
   * 
   * If you are using this property, please join the discussion at proposal [#201](https://github.com/openactive/modelling-opportunity-data/issues/201).
   */
  estimatedDuration?: oa.QuantitativeValue.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * A related video object.
   * 
   * If you are using this property, please join the discussion at proposal [#88](https://github.com/openactive/modelling-opportunity-data/issues/88).
   */
  video?: oa.VideoObject.Type[];
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Internal location of the event, e.g. Court 1
   * 
   * If you are using this property, please join the discussion at proposal [#110](https://github.com/openactive/modelling-opportunity-data/issues/110).
   */
  sportsActivityLocation?: schema.SportsActivityLocation.Type[];
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Duration before the event for which the associated Offers are valid
   * 
   * If you are using this property, please join the discussion at proposal [#204](https://github.com/openactive/modelling-opportunity-data/issues/204).
   */
  offerValidityPeriod?: string;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Whether the event or facility is indoor or outdoor.
   * 
   * If you are using this property, please join the discussion at proposal [#1](https://github.com/openactive/facility-types/issues/1).
   */
  facilitySetting?: oa.FacilitySettingType.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * A property that indicates whether the event is led by a virtual coach. Only relevant if an event `isCoached`. If not provided is assumed to be `false`.
   * 
   * If you are using this property, please join the discussion at proposal [#71](https://github.com/openactive/modelling-opportunity-data/issues/71).
   */
  isVirtuallyCoached?: boolean;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Describes a means of electronic access to a shared virtual space.
   * 
   * If you are using this property, please join the discussion at proposal [#224](https://github.com/openactive/modelling-opportunity-data/issues/224).
   */
  virtualLocation?: oa.VirtualLocation.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * The physical location affiliated with the virtual event, for example the original location of the event before it was moved online.
   * 
   * If you are using this property, please join the discussion at proposal [#227](https://github.com/openactive/modelling-opportunity-data/issues/227).
   */
  affiliatedLocation?: oa.Place.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Indicates whether the virtual event is interactive (e.g. Zoom with participant microphones and cameras on), or is just a one-way broadcast (e.g. Facebook Live, Instagram Live, Zoom with participant microphones and cameras off).
   * 
   * If you are using this property, please join the discussion at proposal [#230](https://github.com/openactive/modelling-opportunity-data/issues/230).
   */
  isInteractivityPreferred?: boolean;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Indicates whether the participant must or may supply equipment for use in the Event.
   * 
   * If you are using this property, please join the discussion at proposal [#229](https://github.com/openactive/modelling-opportunity-data/issues/229).
   */
  participantSuppliedEquipment?: oa.RequiredStatusType.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * The URL of the webpage where the activity provider accepts donations.
   * 
   * If you are using this property, please join the discussion at proposal [#234](https://github.com/openactive/modelling-opportunity-data/issues/234).
   */
  donationPaymentUrl?: string;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * A property that indicates whether the first session is free.
   * 
   * If you are using this property, please join the discussion at proposal [#232](https://github.com/openactive/modelling-opportunity-data/issues/232).
   */
  isFirstSessionAccessibleForFree?: boolean;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * Contact details for an Event, where they are not specifically related to the `organizer` or `leader`.
   * 
   * If you are using this property, please join the discussion at proposal [#113](https://github.com/openactive/modelling-opportunity-data/issues/113).
   */
  contactPoint?: schema.ContactPoint.Type;
  /**
   * [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]
   * The channels through which a booking can be made.
   * 
   * If you are using this property, please join the discussion at proposal [#161](https://github.com/openactive/modelling-opportunity-data/issues/161).
   */
  bookingChannel?: oa.BookingChannelType.Type[];
  /**
   * [NOTICE: This property is part of the Open Booking API Test Interface, and MUST NOT be used in production.]
   */
  testOpenBookingFlow?: oa.TestOpenBookingFlowEnumeration.Type;
  /**
   * [NOTICE: This property is part of the Open Booking API Test Interface, and MUST NOT be used in production.]
   * The opportunity criteria which the Event conforms to.
   */
  testOpportunityCriteria?: oa.TestOpportunityCriteriaEnumeration.Type;
}

/* unfortunately, it must be given this generic type `ObjectSchema<any>` because these schemas can sometimes be
recursively defined, which causes a type error due to potentially infinite nesting of generic params */
/**
 * Event - Validation schema (w/ JOI)
 *
 * This type is derived from https://schema.org/Event, which means that any of this type's properties within schema.org may also be used.
 */
export const JoiSchema = Joi.object({
  '@type': Joi.string().valid('Event').required(),
  identifier: Joi.alternatives().try(Joi.number().integer(), Joi.string(), Joi.lazy(() => oa.PropertyValue.JoiSchema), Joi.array().items(Joi.lazy(() => oa.PropertyValue.JoiSchema))),
  name: Joi.string(),
  description: Joi.string(),
  '@id': Joi.string().uri(),
  accessibilityInformation: Joi.string(),
  accessibilitySupport: Joi.array().items(Joi.lazy(() => oa.Concept.JoiSchema)),
  activity: Joi.array().items(Joi.lazy(() => oa.Concept.JoiSchema)),
  additionalAdmissionRestriction: Joi.array().items(Joi.string()),
  ageRange: Joi.lazy(() => oa.QuantitativeValue.JoiSchema),
  ageRestriction: Joi.lazy(() => oa.QuantitativeValue.JoiSchema),
  attendeeInstructions: Joi.string(),
  category: Joi.alternatives().try(Joi.array().items(Joi.string()), Joi.array().items(Joi.lazy(() => oa.Concept.JoiSchema))),
  contributor: Joi.array().items(Joi.lazy(() => oa.Person.JoiSchema)),
  duration: Joi.string(),
  eventAttendanceMode: Joi.lazy(() => schema.EventAttendanceModeEnumeration.JoiSchema),
  eventStatus: Joi.lazy(() => schema.EventStatusType.JoiSchema),
  genderRestriction: Joi.lazy(() => oa.GenderRestrictionType.JoiSchema),
  image: Joi.array().items(Joi.lazy(() => oa.ImageObject.JoiSchema)),
  isAccessibleForFree: Joi.boolean(),
  isCoached: Joi.boolean(),
  leader: Joi.array().items(Joi.lazy(() => oa.Person.JoiSchema)),
  level: Joi.alternatives().try(Joi.array().items(Joi.string()), Joi.array().items(Joi.lazy(() => oa.Concept.JoiSchema))),
  location: Joi.lazy(() => oa.Place.JoiSchema),
  maximumAttendeeCapacity: Joi.number().integer(),
  maximumVirtualAttendeeCapacity: Joi.number().integer(),
  meetingPoint: Joi.string(),
  offers: Joi.array().items(Joi.lazy(() => oa.Offer.JoiSchema)),
  organizer: Joi.alternatives().try(Joi.lazy(() => oa.Person.JoiSchema), Joi.lazy(() => oa.Organization.JoiSchema)),
  programme: Joi.lazy(() => oa.Brand.JoiSchema),
  remainingAttendeeCapacity: Joi.number().integer(),
  schedulingNote: Joi.string(),
  startDate: Joi.string().isoDate(),
  endDate: Joi.string().isoDate(),
  subEvent: Joi.array().items(Joi.lazy(() => oa.Event.JoiSchema)),
  superEvent: Joi.lazy(() => oa.Event.JoiSchema),
  url: Joi.string().uri(),
  formattedDescription: Joi.string(),
  distance: Joi.lazy(() => oa.QuantitativeValue.JoiSchema),
  attendeeCount: Joi.number().integer(),
  registrationCount: Joi.number().integer(),
  isWheelchairAccessible: Joi.boolean(),
  estimatedDuration: Joi.lazy(() => oa.QuantitativeValue.JoiSchema),
  video: Joi.array().items(Joi.lazy(() => oa.VideoObject.JoiSchema)),
  sportsActivityLocation: Joi.array().items(Joi.lazy(() => schema.SportsActivityLocation.JoiSchema)),
  offerValidityPeriod: Joi.string(),
  facilitySetting: Joi.lazy(() => oa.FacilitySettingType.JoiSchema),
  isVirtuallyCoached: Joi.boolean(),
  virtualLocation: Joi.lazy(() => oa.VirtualLocation.JoiSchema),
  affiliatedLocation: Joi.lazy(() => oa.Place.JoiSchema),
  isInteractivityPreferred: Joi.boolean(),
  participantSuppliedEquipment: Joi.lazy(() => oa.RequiredStatusType.JoiSchema),
  donationPaymentUrl: Joi.string().uri(),
  isFirstSessionAccessibleForFree: Joi.boolean(),
  contactPoint: Joi.lazy(() => schema.ContactPoint.JoiSchema),
  bookingChannel: Joi.array().items(Joi.lazy(() => oa.BookingChannelType.JoiSchema)),
  testOpenBookingFlow: Joi.lazy(() => oa.TestOpenBookingFlowEnumeration.JoiSchema),
  testOpportunityCriteria: Joi.lazy(() => oa.TestOpportunityCriteriaEnumeration.JoiSchema),
});

/**
 * Runtime validator for Event.
 *
 * If some data has a structure which matches a Event, it will be returned with the correct type.
 * Otherwise, this function returns an OAValidationError with details about why the data does not match.
 *
 * Use this to e.g. check a JSON object received in an HTTP request. Example usage (for an express request handler):
 *
 * ```ts
 * const maybeEvent = Event.validate(req.body); // `req.body` will have type `any` or `unknown`
 * if (maybeEvent instanceof OaValidationError) {
 *   // From this point on, `maybeEvent` will have type `OaValidationError`
 *   const error = maybeEvent;
 *   // Do something with the error. Maybe ignore it? Or log it? Or throw? Up to you.
 * }
 * // From this point on, `maybeEvent` will have type `Event.Type`
 * const event = maybeEvent;
 * ```
 */
export function validate(maybeEvent: unknown): Type | OaValidationError {
  const { value, error } = JoiSchema.validate(maybeEvent);
  if (error) {
    return new OaValidationError('Event', maybeEvent, error);
  }
  /* Joi does not implement TS Type Guards, so TS does not implicitly know that this has now been validated
  to have the right type. Therefore, we just cast it to `any`. */
  return value as any;
}
```

File: src/generators/typescript/index.js [hunk 1]
```diff
@@ -43,22 +43,34 @@ class TypeScript extends Generator {
...
+   * @param {string} typeName
+   */
+  filterModel(typeName, model) {
+    /* EventStatusType is erroneously made into a model as well as an enum. Because the TS generator puts enums and
+    models into the same directory, this causes issues */
+    if (typeName === 'schema:EventStatusType') {
+      return false;
+    }
+    return true;
+  }
```
Comment (lukehesluke) [inline src/generators/typescript/index.js:61]: This is unfortunately necessary as it seems that models-lib produces a schema EventStatusType model (which is empty) as well as the enum. This seems like a bug to me. In this code, I've just filtered out the model

You can see an example of this in the existing models-ts project: https://github.com/openactive/models-ts/blob/main/src/schema/models/EventStatusType.js vs https://github.com/openactive/models-ts/blob/main/src/schema/enums/EventStatusType.js

File: tmp/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+src/oa
+src/schema
+lib/
```
Comment (lukehesluke) [inline tmp/.gitignore:3]: I'll remove these files from the `.gitignore` when moving this stuff over to models-ts

File: src/generators/typescript/index.js (hunk 1 above)
Comment (nickevansuk) [reply src/generators/typescript/index.js:61]: This is a very strange bug - doesn't seem to relate to the schema.org model either!

**Comment by nickevansuk** [comment]
> 🚨 Until this is done, validating e.g. a ScheduledSession with superEvent as a SessionSeries will fail, as Joi will say "this is not an Event". See tmp/src/examples/event-inheritance-validation-examples.ts

This seems like a pretty major issue that will need to be resolved before the library is usable?

> Joi does not implement TS Type Guards, so TS does not implicitly know that this has now been validated to have the right type. Therefore, we just cast it to `any`
  
Could we not cast to the type the function is validating for? As for OaValidationError that would get returned separately? Either the result will be a valid of the former, or invalid and then the latter?
  
> [NOTICE: This is a beta property, and is highly likely to change in future versions of this library.]

For beta properties, do we need to actually include the namespace here (e.g. `'beta:registrationCount'`) as JS doesn't have any additional serialisation logic, so it won't know how to output the namespace otherwise?

**Comment by nickevansuk** [comment]
As additional note: in terms of testing this thoroughly before we give it to the implementers, could we write a quick test that just iterates through each of these files: https://github.com/openactive/data-models/tree/master/versions/2.x/examples and both attempts to validate them as input using joi, and creates an example JS file where e.g. a JSdoc annotated const is set to the value of the JSON in the example file, and then the file is validated with typescript?

**Comment by lukehesluke** [comment]
> Could we not cast to the type the function is validating for? As for OaValidationError that would get returned separately? Either the result will be a valid of the former, or invalid and then the latter?

@nickevansuk I'm not sure what you're saying here. OaValidationError is already getting returned separately. What is the issue that you're suggesting a fix for? The function signature of the generated `validate(..)` function is already correct. The code comment is just a reference to the fact that a type assertion has to happen because JOI always returns values as `unknown`s

**Comment by nickevansuk** [comment]
Makes sense to me!

**Pull request merged**